libbpf-tools/bashreadline.bpf.c

```diff
@@ -0,0 +1,35 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (c) 2020 Wenbo Zhang
+#include <vmlinux.h>
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_tracing.h>
+#include "bashreadline.h"
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERF_EVENT_ARRAY);
+	__uint(key_size, sizeof(__u32));
+	__uint(value_size, sizeof(__u32));
+} events SEC(".maps");
+
+SEC("uretprobe/readline")
+int printret(struct pt_regs *ctx) {
+	struct str_t data  = {};
+	char comm[TASK_COMM_LEN] = {};
+	u32 pid;
+	if (!PT_REGS_RC(ctx))
+		return 0;
+	pid = bpf_get_current_pid_tgid() >> 32;
+	data.pid = pid;
+	bpf_probe_read_user(&data.str, sizeof(data.str) - 1, (void *)PT_REGS_RC(ctx));
```

Comment:
should be sizeof(data.str), bpf_probe_read_user() will properly zero-terminate

Reply:
What if we want to read a block of binary?  Use another API?

Reply:
oh, wait, I missed that this is bpf_probe_read, not bpf_probe_read_str(). Shouldn't bpf_probe_read_user_str() be used here?